polaris-service/src/main/java/org/apache/polaris/service/config/DynamicFeatureConfigResolver.java

```diff
@@ -0,0 +1,39 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.polaris.service.config;
+
+import com.fasterxml.jackson.annotation.JsonTypeInfo;
+import io.dropwizard.jackson.Discoverable;
+import java.util.Optional;
+
+/**
+ * DynamicFeatureConfigResolvers dynamically resolve featureConfigurations. This is useful for
+ * integration with feature flag systems which are intended for fetching configs at runtime.
+ */
+@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.PROPERTY, property = "type")
+public interface DynamicFeatureConfigResolver extends Discoverable {
```

Comment:
TBH, I feel that this type _should_ be unnecessary. Unfortunately, we just added the `featureConfiguration` field as a map, rather than having the `DefaultConfigurationStore` instantiated and populated by Jackson during application startup. Now the configuration can't really change to add, e.g., a `type` configuration key so that we could use different `PolarisConfigurationStore` implementations. The `PolarisConfigurationStore` interface was always intended to support dynamic implementations - which is why the `PolarisCallContext` is part of the signature.

I'm just trying to figure out if there isn't a way to make the implementation swappable, while still supporting the backward-compatible configuration. I mean, there's no real harm in adding another interface, but it just feels unnecessary and the consequence of poor decision making earlier on. I hate the idea of that poor decision making having to follow us around forever.

Reply:
Another option would be to make the whole implementation swappable as you mentioned, which could be implemented in a backward compatible way: if a custom type is specified, use it, otherwise use the default. I don't think this change makes that impossible in the future. In fact, you can simply make DynamicFeatureConfigResolver implement the entire configuration store.